### PR TITLE
[luci/service] Infer dynamic shape for constant reshape operation

### PR DESCRIPTION
**Pull request by jongwonyang**
This supports dynamic shape inference for Reshape operation when the new shape is constant.

This PR contains:
- Dynamic output shape inference when the new shape is provided by the node's input and is constant.
- Dynamic output shape inference when the new shape is provided by the node's attribute and is constant.
- Change the location of reshape's shape inference algorithm from `ShapeInferenceRule` to `sinf::Algorithm`.

Related: https://github.com/Samsung/ONE/issues/13697
Draft: https://github.com/Samsung/ONE/pull/13866

Handling non-const value of new shape provided by node's input or attribute is TODO.
I'm considering make `CircleReshape` node to `VariadicArityNode` for this TODO. (like tensorflow lite style)

ONE-DCO-1.0-Signed-off-by: <yjw963@gmail.com>

### CODE REVIEW

File: compiler/luci/import/src/Nodes/CircleReshape.cpp [hunk 1]
```diff
@@ -19,6 +19,8 @@
 #include <luci/IR/Nodes/CircleConst.h>
 #include <luci/IR/Nodes/CircleReshape.h>
 
+#include <iostream>
+
```
Comment (seanshpark) [inline compiler/luci/import/src/Nodes/CircleReshape.cpp:23]: why is this added?

File: compiler/luci/service/src/CircleShapeInference.cpp [hunk 2]
```diff
@@ -28,23 +28,6 @@
 namespace
 {
 
-std::ostream &operator<<(std::ostream &os, const loco::TensorShape &tensor_shape)
```
Comment (seanshpark) [inline compiler/luci/service/src/CircleShapeInference.cpp:31]: why is this removed ?

File: compiler/luci/service/src/CircleShapeInferenceHelper.cpp [hunk 3]
```diff
@@ -161,6 +162,40 @@ loco::TensorShape broadcast_shape(const loco::TensorShape &x, const loco::Tensor
   return output_shape;
 }
 
+loco::TensorShape own_shape(const luci::CircleNode *node)
```
Comment (seanshpark) [inline compiler/luci/service/src/CircleShapeInferenceHelper.cpp:165]: I don't think copying this method here is a good idea.

**Comment by seanshpark** [comment]
title is a bit too long. plz make it shorter to single line.

File: compiler/luci/service/src/Nodes/CircleReshape.test.cpp [hunk 4]
```diff
@@ -37,3 +38,241 @@ TEST(CloneNodeTest, clone_Reshape)
...
+
+  ASSERT_TRUE(shape_inf_rule.infer(&reshape, shape));
+
+  ASSERT_EQ(3, input.rank());
+  ASSERT_TRUE(input.dim(0).known());
+  ASSERT_TRUE(input.dim(1).known());
+  ASSERT_TRUE(input.dim(2).known());
+  ASSERT_EQ(2, input.dim(0).value());
+  ASSERT_EQ(3, input.dim(1).value());
+  ASSERT_EQ(4, input.dim(2).value());
```
Comment (seanshpark) [inline compiler/luci/service/src/Nodes/CircleReshape.test.cpp:70]: shape inference is to infer output shape from inputs.
I don't think we need to check input shape info.
Input shape check disturbs read the test codes.

File: compiler/luci/service/src/Nodes/CircleReshape.test.cpp [hunk 5]
```diff
@@ -37,3 +38,241 @@ TEST(CloneNodeTest, clone_Reshape)
...
+  new_shape.shape_status(luci::ShapeStatus::VALID);
+
+  luci::CircleReshape reshape;
+  reshape.tensor(&input);
+  reshape.shape(&new_shape);
+
+  loco::TensorShape shape;
+  luci::sinf::Rule shape_inf_rule;
+
+  ASSERT_ANY_THROW(shape_inf_rule.infer(&reshape, shape));
```
Comment (seanshpark) [inline compiler/luci/service/src/Nodes/CircleReshape.test.cpp:208]: why is it thrown?

File: compiler/luci/service/src/Nodes/CircleReshape.cpp [hunk 6]
```diff
@@ -34,4 +40,110 @@ luci::CircleNode *CloneNodeLet<CN::OPQR>::visit(const luci::CircleReshape *node)
...
+  if (!(shape_by_input == shape_by_attr))
+  {
+    INFO(l) << "CircleReshape: Two new shape information mismatched : " << std::endl;
+    INFO(l) << "   shape_by_input : " << shape_by_input << std::endl;
+    INFO(l) << "   shape_by_attr : " << shape_by_attr << std::endl;
+  }
+
+  loco::TensorShape output_shape = shape_by_input;
+
+  // One of the dimensions can have special value -1, meaning its actual value should be inferred.
```
Comment (seanshpark) [inline compiler/luci/service/src/Nodes/CircleReshape.cpp:116]: have you changed any logic here?
existing code logic is based with infer with known shape and do infer as much as possible with unknown shapes.

title if this PR is to do shape infer with dynamic shape.
I think this logic should be changed.
please discuss with your SAFFY mantor about how to.

File: compiler/luci/service/src/Nodes/CircleReshape.cpp [hunk 7]
```diff
@@ -34,4 +40,110 @@ luci::CircleNode *CloneNodeLet<CN::OPQR>::visit(const luci::CircleReshape *node)
...
+      {
+        shape_by_input.dim(axis) = const_shape_node->at<S32>(axis);
+        if (const_shape_node->at<S32>(axis) < 0)
+          shape_by_input.dim(axis).unset();
+      }
+    }
+    else
+    {
+      // We use shape from the node itself
+      shape_by_input = own_shape(node);
```
Comment (llFreetimell) [inline compiler/luci/service/src/Nodes/CircleReshape.cpp:93]: As I said in https://github.com/Samsung/ONE/pull/13866#discussion_r1739818477, these lines should be substitued by `shape_by_attr` logic.

File: compiler/luci/service/src/Nodes/CircleReshape.cpp [hunk 8]
```diff
@@ -34,4 +40,110 @@ luci::CircleNode *CloneNodeLet<CN::OPQR>::visit(const luci::CircleReshape *node)
...
+    INFO(l) << "   shape_by_input : " << shape_by_input << std::endl;
+    INFO(l) << "   shape_by_attr : " << shape_by_attr << std::endl;
+  }
+
+  loco::TensorShape output_shape = shape_by_input;
+
+  // One of the dimensions can have special value -1, meaning its actual value should be inferred.
+  if (is_input_known)
+  {
+    const auto input_shape = luci::shape_get(node->tensor()).as<loco::TensorShape>();
```
Comment (llFreetimell) [inline compiler/luci/service/src/Nodes/CircleReshape.cpp:119]:
```suggestion
    const auto input_shape = circle_shape(node->tensor());
```

File: compiler/luci/service/src/Nodes/CircleReshape.cpp [hunk 9]
```diff
@@ -34,4 +40,110 @@ luci::CircleNode *CloneNodeLet<CN::OPQR>::visit(const luci::CircleReshape *node)
...
+    uint32_t input_element_count = 1;
+    uint32_t output_element_count = 1;
+    uint32_t unknown_dim_index = UINT32_MAX;
+    for (uint32_t i = 0; i < input_shape.rank(); ++i)
+      input_element_count *= (input_shape.dim(i).known() ? input_shape.dim(i).value() : 1);
```
Comment (llFreetimell) [inline compiler/luci/service/src/Nodes/CircleReshape.cpp:124]: If all dimensions of input node are known, we do not need to check known/unknown.

```suggestion
      input_element_count *= input_shape.dim(i).value();
```

File: compiler/luci/service/src/Nodes/CircleReshape.cpp [hunk 10]
```diff
@@ -34,4 +40,110 @@ luci::CircleNode *CloneNodeLet<CN::OPQR>::visit(const luci::CircleReshape *node)
...
+      break;
+    }
+  }
+
+  loco::TensorShape shape_by_input;
+  {
+    LUCI_ASSERT(node->shape(), "2nd input shape() should not be nullptr");
+
+    // Only support node's shape() is CircleConst with S32
+    // TODO support other node with other types
```
Comment (llFreetimell) [inline compiler/luci/service/src/Nodes/CircleReshape.cpp:75]:
```suggestion
    // https://github.com/tensorflow/tensorflow/blob/2fe29379f6449d263b787eb636020fd81017b8bf/tensorflow/lite/kernels/reshape.cc#L130
```

File: compiler/luci/import/src/Nodes/CircleReshape.cpp (hunk 1 above)
Comment (jongwonyang) [reply compiler/luci/import/src/Nodes/CircleReshape.cpp:23]: Sorry, I added it for debugging but forgot to remove it. I'll remove it!

File: compiler/luci/service/src/CircleShapeInference.cpp (hunk 2 above)
Comment (jongwonyang) [reply compiler/luci/service/src/CircleShapeInference.cpp:31]: I moved it to `compiler/luci/service/src/CircleShapeInferenceHelper.cpp` because I thought it would be more appropriate place for this.
If you don't think so, I'll rollback this change.

File: compiler/luci/service/src/CircleShapeInferenceHelper.cpp (hunk 3 above)
Comment (jongwonyang) [reply compiler/luci/service/src/CircleShapeInferenceHelper.cpp:165]: > I don't think copying this method here is a good idea.

Thanks for the feedback.
I was migrating shape inference rule from `compiler/luci/service/src/CircleShapeInferenceRule.cpp` to `compiler/luci/service/src/Nodes/CircleReshape.cpp`.
And I found out that `own_shape` is being used in many inference logics including reshape.
So I copied it to `CircleShapeInferenceHelper.cpp`.

I think it could be redundant.
Since it was in a private namespace, I could't use it in `compiler/luci/service/src/Nodes/CircleReshape.cpp`.
Could you give me some advice for how to use `own_shape` in other files?

File: compiler/luci/service/src/Nodes/CircleReshape.test.cpp (hunk 4 above)
Comment (jongwonyang) [reply compiler/luci/service/src/Nodes/CircleReshape.test.cpp:70]: You're right. Thanks for your feedback.
I'll remove all the assertions not related to output shape inference.

File: compiler/luci/service/src/CircleShapeInferenceHelper.cpp (hunk 3 above)
Comment (seanshpark) [reply compiler/luci/service/src/CircleShapeInferenceHelper.cpp:165]: I need to explain about this method.
existing shape inference isn't for dynamic shape, began with only static shape, historically.
but then there appeared models with unknown batch (dim 0 for rank 4 shape).
our codes had to mitigate this situation but anyway our main target needed only static shape,
so we've done some shape information passing from input to output (not intelligent shape inference) or do some minimum shape inference or just reuse the shape information from the file.

this `own_shape()` to just to pass what ever was in the first place when our poor shape inference logic couldn't handle it.

as I looked in to this PR, this PR tells it's doing some smart shape inference (by the comment) and should not use this method if the logic is well read and understood.

anyway, current PR has lots of changes in multiple purposes (as the comment says) that would be better to be separated for single purpose.

FYI @shs-park , @zetwhite

File: compiler/luci/service/src/Nodes/CircleReshape.test.cpp (hunk 5 above)
Comment (jongwonyang) [reply compiler/luci/service/src/Nodes/CircleReshape.test.cpp:208]: I threw an exception when the number of unknown dimensions is more than 1 in new shape.
I follow the behavior of the tensorflow lite's, and I thought it could be reasonable because constant new shape can be inferred only up for 1 dimension.
Tensorflow lite throws following error in that case: `error: 'tf.Reshape' op requires 'shape' to have at most one dynamic dimension, but got multiple dynamic dimensions at indices 0 and 1`

File: compiler/luci/service/src/CircleShapeInferenceHelper.cpp (hunk 3 above)
Comment (jongwonyang) [reply compiler/luci/service/src/CircleShapeInferenceHelper.cpp:165]: Thanks a lot.
I have fully understood about the `own_shape` method.
I'm planning to not to use `own_shape` in my new shape inference algorithm.

And about the current PR, I admit it's very complicated.
Sorry for the confusion. It was my first commitment to an open source project and I had to be more careful.

I think I need to start over my PR from draft with following 3 separate parts:
1. Migrate current reshape algorithm to `sinf::Algorithm`
2. Implement dynamic shape inference algorithm for const inputs.
3. Implement dynamic shape inference algorithm for non-const inputs.

If you don't mind I'll close this PR and start over with my SSAFY mentors :)

Comment (seanshpark) [reply compiler/luci/service/src/CircleShapeInferenceHelper.cpp:165]: > If you don't mind I'll close this PR and start over with my SSAFY mentors :)

no problem and it's OK with me.

File: compiler/luci/service/src/Nodes/CircleReshape.cpp (hunk 7 above)
Comment (jongwonyang) [reply compiler/luci/service/src/Nodes/CircleReshape.cpp:93]: Thank you so much for your kind and insightful reviews.
I’ve carefully read through all of them and applied the suggested changes, except for the `own_shape` one.
However, I now have a complete understanding of the `own_shape` method, and I will work on replacing it with new logic (including updating the recipe).

As I mentioned in https://github.com/Samsung/ONE/pull/13912#discussion_r1741534197, I would like to close this PR and start over to ensure a more readable and understandable commit.

I sincerely hope you continue to take an interest in this issue and offer your support. 
Thanks again!

Comment (llFreetimell) [reply compiler/luci/service/src/Nodes/CircleReshape.cpp:93]: Thank you for the kind response and sorry(?) for your effort :)
After creating new PR, please register me as a reviewer!
I will gladly take an interest for that :D
